site/content/blog/2019-11-04-unit-testing-mmctl-commands.md

```diff
@@ -0,0 +1,140 @@
+
+```go
+func (s *MmctlUnitTestSuite) TestSearchUserCmd() {
+    s.Run("Our test case", func() {
+        // ...
+    })
+}
+```
+
+This way, we will have a `s` instance inside our test function that will contain a mocked instance of the client ready for us to use, and we will be able to use the suite to run our assertions.
+
+# Mocking an interaction with the server
+
+So we have our test function defined, next step is to think about our test case: what inputs is the command going to receive? What interactions with the server are those inputs going to cause? And what responses do I want to mock for them?
+
+For our given test case, we are going to receive an email address as the input, and we want to mock a valid response for it. The client method that our command is going to use to send the request to the server is the `GetUserByEmail` method, and for the case of an existing user, it will return the found user instance:
+
+```go
+emailArg := "example@example.com"
+mockUser := model.User{Username: "ExampleUser", Email: emailArg}
+
+s.client.
+	EXPECT().
+	GetUserByEmail(emailArg, "").
+	Return(&mockUser, &model.Response{Error: nil}).
+	Times(1)
+```
+
+First we define our input, the `emailArg` that the command is going to receive, and the `mockUser` we are going to mock as the server's response for the user search request. As the response is valid, we will create a simple response that contains no error.
+
+Then we have to state the expectations of the mock. We expect our command to use the `GetUserByEmail` function, calling it with the `emailArg` we mocked and a second argument that will always be an empty string, and we want the server to return our mocked user as a response for that call. Last, we state that this call should happen once.
+
+During our test run, if the client's method is called with the arguments that we expect, it will respond accordingly. At the end of the test, `gomock` will check that our assertions were correct, and it will mark the test as failed if they were not.
+
+# Asserting the command's behavior
+
+After mocking the server interactions, all that is left to do is running the command and checking the outputs. There are two things that we can use to check that the command behaved as we expected: the command's return value, and whatever was printed during its execution.
+
+To perform this assertions we can use the suite itself for [assertions](https://godoc.org/github.com/stretchr/testify/assert) or `s.Require()` for [requires](https://godoc.org/github.com/stretchr/testify/require). Both implement the same helpers to check that our values meet our expectations, the difference is that a failed check with `require` will mark the test as failed and will stop it right after, and a failed check with `assert` will mark the test as failed too, but it will continue running it.
```

Comment:
```suggestion
To perform this assertion we can use the suite itself for [assertions](https://godoc.org/github.com/stretchr/testify/assert) or `s.Require()` for [requires](https://godoc.org/github.com/stretchr/testify/require). Both implement the same helpers to check that our values meet our expectations, the difference is that a failed check with `require` will mark the test as failed and will stop immediately. A failed check with `assert` will also mark the test as failed, but will continue running.
```

Reply:
I think I'd rather go the other way around and change this to:

```suggestion
To perform these assertions we can use the suite itself for [assertions](https://godoc.org/github.com/stretchr/testify/assert) or `s.Require()` for [requires](https://godoc.org/github.com/stretchr/testify/require). Both implement the same helpers to check that our values meet our expectations, the difference is that a failed check with `require` will mark the test as failed and will stop it right after, and a failed check with `assert` will mark the test as failed too, but it will continue running it.
```

What do you think @justinegeffen?

Reply:
Yeah that works for me; reads much better. :)